<a href="https://colab.research.google.com/github/KeqingW44448/api/blob/main/RSM8421/Assignments/Assignment%20One/Problem_6_Coding_Part_for_Jupyternote_Homework_Assignment_One.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Overview
This assignment consists of several exercises focused on **debugging neural networks**. Each network provided contains at least one issue. Your task is to identify the problem, explain your reasoning, and apply the appropriate debugging techniques you have learned in class to resolve it.

In [ ]:
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from keras.models import Sequential
from keras.layers import Input, Dense, Activation, Dropout
from keras.datasets import mnist
from keras.optimizers import Adam
from keras.regularizers import l1, l2
from keras.utils import to_categorical
from keras.callbacks import TensorBoard
from keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler

## Exercise 6

Let's test how well we can improve the Boston Housing model's test MAE when using only a subset of the data.

In [ ]:
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()

np.random.seed(0)
index = np.arange(len(x_train))

# Take a biased training set - top 50 house prices
index = sorted(index, key=lambda x: -y_train[x])
x_train = x_train[index]
y_train = y_train[index]

sample = 100
x_train, y_train = x_train[:sample], y_train[:sample]
x_test, y_test = x_test, y_test

scalar = StandardScaler()
x_train2 = scalar.fit_transform(x_train)
x_test2 = scalar.transform(x_test)

print("Training Dataset Size: ", x_train2.shape)
print("Training Labels Size: ", y_train.shape)
print("Testing Dataset Size: ", x_test2.shape)
print("Testing Labels Size: ", y_test.shape)

57026/57026 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Training Dataset Size:  (100, 13)
Training Labels Size:  (100,)
Testing Dataset Size:  (102, 13)
Testing Labels Size:  (102,)


In [ ]:
model1 = Sequential()
model1.add(Input(shape=(13,)))
model1.add(Dense(100, activation='relu'))
model1.add(Dense(100, activation='relu'))
model1.add(Dense(100, activation='relu'))
model1.add(Dense(1,))

model1.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │         1,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 100)            │        10,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 100)            │        10,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │           101 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 21,701 (84.77 KB)

 Trainable params: 21,701 (84.77 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
model1.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
callbacks = []
history = model1.fit(x_train2, y_train, batch_size=sample, epochs=100, callbacks=callbacks, verbose=1)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 5s 5s/step - loss: 34.9386
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 82ms/step - loss: 33.2492
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 188ms/step - loss: 30.3603
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 126ms/step - loss: 25.2174
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 123ms/step - loss: 17.6997
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 105ms/step - loss: 10.8340
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 134ms/step - loss: 13.7090
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step - loss: 13.9750
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 88ms/step - loss: 8.4827
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 170ms/step - loss: 5.0649
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - loss: 6.8990
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 178ms/step - loss: 7.7399
Epoch 13/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 124ms/step - loss: 7.0439
Epoch 14/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 63ms/step - loss: 6.1059
Epoch 15/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 85ms/step - loss: 6.3933
Epoch

In [ ]:
mae = model1.evaluate(x_test2, y_test)
print("MAE: %.4f" % mae)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - loss: 17.0352
MAE: 14.5558


## Problem 6

Using the same number of epochs and the same training set, apply regularization (without altering the training set) to improve test set accuracy. Describe the regularization methods you used, including the parameters you tried. Which method worked best, and why?

## Add L2: kernel_regularizer=l2(0.01)

In [ ]:
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from keras.models import Sequential
from keras.layers import Input, Dense, Activation, Dropout
from keras.datasets import mnist
from keras.optimizers import Adam
from keras.regularizers import l1, l2
from keras.utils import to_categorical
from keras.callbacks import TensorBoard
from keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler

In [ ]:
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()

np.random.seed(0)
index = np.arange(len(x_train))

# Take a biased training set - top 50 house prices
index = sorted(index, key=lambda x: -y_train[x])
x_train = x_train[index]
y_train = y_train[index]

sample = 100
x_train, y_train = x_train[:sample], y_train[:sample]
x_test, y_test = x_test, y_test

scalar = StandardScaler()
x_train2 = scalar.fit_transform(x_train)
x_test2 = scalar.transform(x_test)

print("Training Dataset Size: ", x_train2.shape)
print("Training Labels Size: ", y_train.shape)
print("Testing Dataset Size: ", x_test2.shape)
print("Testing Labels Size: ", y_test.shape)

Training Dataset Size:  (100, 13)
Training Labels Size:  (100,)
Testing Dataset Size:  (102, 13)
Testing Labels Size:  (102,)


In [ ]:
from keras.regularizers import l1, l2

# Example with L2 regularization
model_l2 = Sequential([
    Input(shape=(13,)),
    Dense(100, activation='relu', kernel_regularizer=l2(0.01)),
    Dense(100, activation='relu', kernel_regularizer=l2(0.01)),
    Dense(100, activation='relu', kernel_regularizer=l2(0.01)),
    Dense(1)
])

model_l2.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
history_l2 = model_l2.fit(x_train2, y_train, batch_size=sample, epochs=100, verbose=0)
mae_l2 = model_l2.evaluate(x_test2, y_test)
print("L2 MAE:", mae_l2)

model1.summary()


4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 19.8858  
L2 MAE: 17.437114715576172


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │         1,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 100)            │        10,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 100)            │        10,100 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │           101 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 65,105 (254.32 KB)

 Trainable params: 21,701 (84.77 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 43,404 (169.55 KB)

In [ ]:
model1.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
callbacks = []
history = model1.fit(x_train2, y_train, batch_size=sample, epochs=100, callbacks=callbacks, verbose=1)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step - loss: 1.8943
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 42ms/step - loss: 17.9022
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - loss: 8.7130
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 3.6273
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 8.2293
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step - loss: 9.2133
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 48ms/step - loss: 7.5803
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 4.0188
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 44ms/step - loss: 3.4924
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - loss: 5.7900
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 5.7192
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 49ms/step - loss: 3.0813
Epoch 13/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 45ms/step - loss: 2.2370
Epoch 14/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step - loss: 4.0675
Epoch 15/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 47ms/step - loss: 4.1545
Epoch 16/100
1/1 ━━━

In [ ]:
mae = model1.evaluate(x_test2, y_test)
print("MAE: %.4f" % mae)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 15.5639  
MAE: 13.7793


# Second Step: Regulization: L1 (|x|) and L2 (x^2)
##Add L1 & L2 Regularization Grid Search with alpha values: [0.0001, 0.001, 0.01, 0.1] to select the optimal mae_test

In [ ]:
from tensorflow import keras
import numpy as np
from keras.models import Sequential
from keras.layers import Input, Dense
from keras.optimizers import Adam
from keras.regularizers import l1, l2
from keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler

# -----------------------------
# Load and preprocess data
# -----------------------------
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()

np.random.seed(0)
index = np.arange(len(x_train))

# Take a biased training set - top 100 house prices
index = sorted(index, key=lambda x: -y_train[x])
x_train = x_train[index]
y_train = y_train[index]

sample = 100
x_train, y_train = x_train[:sample], y_train[:sample]

# Scale features
scaler = StandardScaler()
x_train2 = scaler.fit_transform(x_train)
x_test2 = scaler.transform(x_test)

print("Training Dataset Size:", x_train2.shape)
print("Testing Dataset Size:", x_test2.shape)

# -----------------------------
# Grid search for L1 and L2 alpha
# -----------------------------
alphas = [1e-4, 1e-3, 1e-2, 0.1]
results_l1 = {}
results_l2 = {}

for alpha in alphas:
    # L1 model
    model_l1 = Sequential([
        Input(shape=(13,)),
        Dense(100, activation='relu', kernel_regularizer=l1(alpha)),
        Dense(100, activation='relu', kernel_regularizer=l1(alpha)),
        Dense(100, activation='relu', kernel_regularizer=l1(alpha)),
        Dense(1)
    ])
    model_l1.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
    model_l1.fit(x_train2, y_train, batch_size=sample, epochs=100, verbose=0)
    mae_l1 = model_l1.evaluate(x_test2, y_test, verbose=0)
    results_l1[alpha] = mae_l1
    print(f"L1 alpha={alpha}: MAE={mae_l1:.4f}")

    # L2 model
    model_l2 = Sequential([
        Input(shape=(13,)),
        Dense(100, activation='relu', kernel_regularizer=l2(alpha)),
        Dense(100, activation='relu', kernel_regularizer=l2(alpha)),
        Dense(100, activation='relu', kernel_regularizer=l2(alpha)),
        Dense(1)
    ])
    model_l2.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
    model_l2.fit(x_train2, y_train, batch_size=sample, epochs=100, verbose=0)
    mae_l2 = model_l2.evaluate(x_test2, y_test, verbose=0)
    results_l2[alpha] = mae_l2
    print(f"L2 alpha={alpha}: MAE={mae_l2:.4f}")


# Find best alpha for L1 and L2

best_alpha_l1 = min(results_l1, key=results_l1.get)
best_alpha_l2 = min(results_l2, key=results_l2.get)

print(f"\nBest L1 alpha: {best_alpha_l1} with MAE={results_l1[best_alpha_l1]:.4f}")
print(f"Best L2 alpha: {best_alpha_l2} with MAE={results_l2[best_alpha_l2]:.4f}")

Training Dataset Size: (100, 13)
Testing Dataset Size: (102, 13)
L1 alpha=0.0001: MAE=16.8601
L2 alpha=0.0001: MAE=18.4757
L1 alpha=0.001: MAE=15.7742
L2 alpha=0.001: MAE=17.5671
L1 alpha=0.01: MAE=13.3652
L2 alpha=0.01: MAE=14.0781
L1 alpha=0.1: MAE=12.8791
L2 alpha=0.1: MAE=10.0345

Best L1 alpha: 0.1 with MAE=12.8791
Best L2 alpha: 0.1 with MAE=10.0345


## Conclusion: The regulization for L2-parameter regulization with the weight alpha = 0.1 is the optimal hyperparameter structure with the minimum MAE = 10.0345.

 Explanation: Using the same one hundred epochs and the same training set, apply regularization (without altering the training set) to improve test set accuracy. The regularization methods I used, including the parameteres I tried are L1 - and L2 - Parameter Regularization with alphas = [1e-4, 1e-3, 1e-2, 0.1].

L2 regularization works best because:

Boston Housing dataset has correlated features and continuous values.
L2 reduces overfitting without making weights sparse (which L1 does).
Dropout is less effective for tabular data with small sample size.

## Final Step: Dropout Regularization Grid Search with p = 0.1, 0.3, 0.5, 0.7

In [ ]:
from tensorflow import keras
import numpy as np
from keras.models import Sequential
from keras.layers import Input, Dense, Dropout
from keras.optimizers import Adam
from keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler


# Load and preprocess data
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()

np.random.seed(0)
index = np.arange(len(x_train))

# Take a biased training set - top 100 house prices
index = sorted(index, key=lambda x: -y_train[x])
x_train = x_train[index]
y_train = y_train[index]

sample = 100
x_train, y_train = x_train[:sample], y_train[:sample]

# Scale features
scaler = StandardScaler()
x_train2 = scaler.fit_transform(x_train)
x_test2 = scaler.transform(x_test)

print("Training Dataset Size:", x_train2.shape)
print("Testing Dataset Size:", x_test2.shape)

# Grid search for Dropout rate
dropout_rates = [0.1, 0.3, 0.5, 0.7]
results = {}

for rate in dropout_rates:
    model = Sequential([
        Input(shape=(13,)),
        Dense(100, activation='relu'),
        Dropout(rate),
        Dense(100, activation='relu'),
        Dropout(rate),
        Dense(100, activation='relu'),
        Dropout(rate),
        Dense(1)
    ])

    model.compile(loss='mae', optimizer=Adam(learning_rate=0.01))
    model.fit(x_train2, y_train, batch_size=sample, epochs=100, verbose=0)
    mae = model.evaluate(x_test2, y_test, verbose=0)
    results[rate] = mae
    print(f"Dropout rate={rate}: MAE={mae:.4f}")

# Find best dropout rate
best_rate = min(results, key=results.get)
print(f"\nBest Dropout rate: {best_rate} with MAE={results[best_rate]:.4f}")

Training Dataset Size: (100, 13)
Testing Dataset Size: (102, 13)
Dropout rate=0.1: MAE=15.5760
Dropout rate=0.3: MAE=17.8819
Dropout rate=0.5: MAE=11.6088
Dropout rate=0.7: MAE=9.9479

Best Dropout rate: 0.7 with MAE=9.9479


## Conclusion: The regulization for dropout randomly ignore fraction p = 0.7 of neurons is the optimal hyperparameter structure with the minimum MAE = 9.9479

 Explanation: Using the same one hundred epochs and the same training set, apply regularization (without altering the training set) to improve test set accuracy. The regularization methods I used, including the parameteres I tried are the dropout with p = [0.1, 0.3, 0.5, 0.7].

Drop out works the best here, as the sample size is sufficicently large and the numbers of the important features and attributions are very large, therefore, the drop out will perform as the optimal result with test MSE = 9.9479.